### ISSUE TITLE
JDK15 jdk_foreign_0_FAILED : TEST: java/foreign/TestMismatch.java

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse/openj9#10580]
Failure link
------------

From an internal build `Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/58`:
```
22:19:15  openjdk version "15-internal" 2020-09-15
22:19:15  OpenJDK Runtime Environment (build 15-internal+0-adhoc.jenkins.BuildJDK15ppc64lelinuxNightly)
22:19:15  Eclipse OpenJ9 VM (build master-5ef8df8f481, JRE 15 Linux ppc64le-64-Bit Compressed References 20200910_73 (JIT enabled, AOT enabled)
22:19:15  OpenJ9   - 5ef8df8f481
22:19:15  OMR      - 58b0c18383a
22:19:15  JCL      - ad829e12c8f based on jdk-15+36)
```
For example, to rebuild the failed tests in <Jenkins URL>=https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder, use the following links:
00:20:53  https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?JDK_VERSION=15&JDK_IMPL=openj9&BUILD_LIST=openjdk&PLATFORM=ppc64le_linux&TARGET=jdk_foreign_0

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------

```
00:14:21  --------------------------------------------------
00:14:21  TEST: java/foreign/TestMismatch.java
00:14:21  TEST JDK: /home/jenkins/workspace/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/openjdkbinary/j2sdk-image

00:14:21  STDOUT:
00:14:21  [TestNG] Running:
00:14:21    java/foreign/TestMismatch.java
00:14:21  
00:14:21  test TestMismatch.testClosed(): success
00:14:21  testDifferentValues s1:MemorySegment{ id=0x51cdef15 limit: 16 }, s2:MemorySegment{ id=0x512432c5 limit: 16 }
00:14:21  test TestMismatch.testDifferentValues(jdk.internal.foreign.NativeMemorySegmentImpl@31801a63, jdk.internal.foreign.NativeMemorySegmentImpl@778f06b5): success

00:14:21  ...
00:14:21  Output overflow:
00:14:21  JT Harness has limited the test output to the text
00:14:21  at the beginning and the end, so that you can see how the
00:14:21  test began, and how it completed.
00:14:21  

00:14:21  test TestMismatch.testDifferentValues(jdk.internal.foreign.HeapMemorySegmentImpl@369a25a0, jdk.internal.foreign.HeapMemorySegmentImpl@63b04c33): success
00:14:21  testDifferentValues s1:MemorySegment{ id=0x3bcc79bd limit: 1 }, s2:MemorySegment{ id=
00:14:21  TEST RESULT: Error. Agent error: java.lang.Exception: Agent 2 timed out with a timeout of 960 seconds; check console log for any additional details
00:14:21  --------------------------------------------------
00:14:21  Test results: passed: 16; error: 1
00:14:59  Report written to /home/jenkins/workspace/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/jvmtest/openjdk/report/html/report.html
00:14:59  Results written to /home/jenkins/workspace/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/openjdk-tests/TKG/test_output_15997909705174/jdk_foreign_0/work
00:14:59  Error: Some tests failed or other problems occurred.
00:14:59  
00:14:59  jdk_foreign_0_FAILED
```

**Comment by liqunl** [comment]
On the xlinuxXL failure from #10588 has the same error message. When timeout happened, the program ran jstack and printed the following information
```
09:46:32  "AgentVMThread" prio=5 Id=26 RUNNABLE
09:46:32  	at java.base@15/jdk.internal.util.ArraysSupport.vectorizedMismatch(ArraysSupport.java:124)
09:46:32  	at java.base@15/jdk.internal.util.ArraysSupport.vectorizedMismatchLargeForBytes(ArraysSupport.java:180)
09:46:32  	at jdk.incubator.foreign@15/jdk.internal.foreign.AbstractMemorySegmentImpl.mismatch(AbstractMemorySegmentImpl.java:155)
09:46:32  	at app//TestMismatch.testLargeAcrossMaxBoundary(TestMismatch.java:131)
09:46:32  	at app//TestMismatch.testLarge(TestMismatch.java:120)
```

This is the agent that timed out. `vectorizedMismatch` is annotated as HotSpotIntrinsicCandidate and judging from the names of the first two methods on the stack, they're supposed to be vectorized?

If Openj9 doesn't optimize the method, we'll run slower. @andrewcraik @DanHeidinga Should we do the same optimization, or just ignore the timeout?

**Comment by andrewcraik** [comment]
@liqunl if we increase the timeout does the test pass? I don't know that performance should be a blocker for a release we should just carry a perf item into the next release to address it...

**Comment by liqunl** [comment]
Adding 2 grinders from @andrew-m-leonard, I believe it's two iterations, both have passed
https://ci.adoptopenjdk.net/job/Grinder/3859
https://ci.adoptopenjdk.net/job/Grinder/3865

A 20-iteration grinder has also passed.
https://ci.adoptopenjdk.net/job/Grinder/3867

**Comment by andrewcraik** [comment]
@liqunl is this evidence that this is a performance issue?

**Comment by liqunl** [comment]
No. We haven't ruled out the possibility of a hang. The timeout is 960 seconds and it should be long enough to finish the test. I'm trying to reproduce it in grinder and also locally.

**Comment by liqunl** [comment]
This grinder is closer to the original one, runs on x86_linux_xl. The running time of the whole foreign test group is around 5 minutes.
100 iterations
https://ci.adoptopenjdk.net/job/Grinder/3875/
https://ci.adoptopenjdk.net/job/Grinder/3876/
https://ci.adoptopenjdk.net/job/Grinder/3877/
https://ci.adoptopenjdk.net/job/Grinder/3878/
https://ci.adoptopenjdk.net/job/Grinder/3879/

**Comment by liqunl** [comment]
Only the 3875 grinder finished, the rest failed to download the jdk due to bad jdk url.
80 iterations
https://ci.adoptopenjdk.net/job/Grinder/3886/
https://ci.adoptopenjdk.net/job/Grinder/3887/
https://ci.adoptopenjdk.net/job/Grinder/3888/
https://ci.adoptopenjdk.net/job/Grinder/3889/

**Comment by liqunl** [comment]
So far, 100 iterations have passed, another 100 iterations
https://ci.adoptopenjdk.net/job/Grinder/3891/
https://ci.adoptopenjdk.net/job/Grinder/3892/
https://ci.adoptopenjdk.net/job/Grinder/3893/
https://ci.adoptopenjdk.net/job/Grinder/3894/
https://ci.adoptopenjdk.net/job/Grinder/3895/

**Comment by liqunl** [comment]
@andrewcraik Not reproducible in 200 jobsm is it enough to downgrade its severity?

**Comment by liqunl** [comment]
I have another 80 iter grinder passed
https://ci.adoptopenjdk.net/job/Grinder/3924/
https://ci.adoptopenjdk.net/job/Grinder/3925/
https://ci.adoptopenjdk.net/job/Grinder/3926/
https://ci.adoptopenjdk.net/job/Grinder/3927/

300 jobs have passed in total, not able to reproduce. Since this issue is not a stop-ship for 0.22 release and it shouldn't be based on the low failure rate.

**Comment by andrewcraik** [comment]
The non-crash issue has not reproduced in 300 jobs. Perhaps it was timing if it ran on a shared machine and resources were low? I think we need to wait for a reproduction in the hope we can get some more information from that.

**Comment by pshipton** [comment]
Have you tried grinding it on the internal machines where the original problem occurred?

**Comment by liqunl** [comment]
> Have you tried grinding it on the internal machines where the original problem occurred?

Not yet, let me try

**Comment by liqunl** [comment]
https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/10355
So 20-job grinder on original job on ppcle showed that timeout happened 3/20, but all happened with a crash in agent. So the timeout on ppcle is closed related to the crash, should be the same issue to #10588. The timeout on X is different, more random, could be due to timing or resource constraint, and is not reproducible in 300 jobs.

I believe when the crash is fixed, the timeout on ppcle will go away.

**Comment by pshipton** [comment]
Let's close this as a dup of https://github.com/eclipse/openj9/issues/10588

**Comment by pshipton** [comment]
I think we'd better reopen this. #10588 has identified a ppc code gen problem, however this can't be causing the test failure on xlinux.
```
16:20:35  "AgentVMThread" prio=5 Id=26 RUNNABLE
16:20:35  	at java.base@15.0.1-internal/jdk.internal.util.ArraysSupport.vectorizedMismatch(ArraysSupport.java:124)
16:20:35  	at java.base@15.0.1-internal/jdk.internal.util.ArraysSupport.vectorizedMismatchLargeForBytes(ArraysSupport.java:180)
16:20:35  	at jdk.incubator.foreign@15.0.1-internal/jdk.internal.foreign.AbstractMemorySegmentImpl.mismatch(AbstractMemorySegmentImpl.java:155)
16:20:35  	at app//TestMismatch.testLargeAcrossMaxBoundary(TestMismatch.java:132)
```

**Comment by liqunl** [comment]
@pshipton Is this failure more reproducible on the internal build you're seeing? We haven't been able to reproduce in 300 jobs. And we can't figure out anything with just the output. If we can't reproduce it and there's no core dump from original failure, there's nothing we can do.

**Comment by pshipton** [comment]
@liqunl pls take a look at the internal link I provided (job/Test_openjdknext_j9_sanity.openjdk_x86-64_linux_Personal/12). It failed in the last 2 builds, and perhaps others before that.

**Comment by liqunl** [comment]
I'm able to reproduce the timeout consistently on internal machine, and with `-Xint`, but can't reproduce it on my own VM. The difference is that internal machine has only 3G free memory, and my VM has 9G. The test that hangs is
https://github.com/ibmruntimes/openj9-openjdk-jdk15/blob/openj9/test/jdk/java/foreign/TestMismatch.java#L113

It's allocating 2G native memory each segment, 4G in total. 

When it hangs, one process has taken 92% of the system's memory
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 4196 jenkins   20   0 7083660 3.579g    748 S  11.3 92.7   0:25.15 java
   47 root      20   0       0      0      0 S   5.6  0.0   4:43.45 kswapd0
```
kswapd0 is also running swapping the memory.
And the free memory is
```
$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3951        3744         147           0          58          46
Swap:          1999        1235         764
```

In summary, this is a resource constrain issue, we shouldn't run this test on machines with insufficient memory.

**Comment by pshipton** [comment]
Great work. I hate to exclude the test, since it found a bug on plinux. I suppose we could just exclude it on xlinux. I'm not sure how often it fails on the OpenJ9 machines, if ever. Or how often on the Adopt machines. Or if the xlinux failures are always timeouts.

@jdekonin seems the internal machines are light on resources.

**Comment by jdekonin** [comment]
2 cpu, 4G ram system use to be standard test machine size.  Is this not the case any more?  Is there a different minimum that we should be using?

**Comment by pshipton** [comment]
Apparently so, obviously some of the openjdk tests require more.

**Comment by liqunl** [comment]
> Great work. I hate to exclude the test, since it found a bug on plinux. I suppose we could just exclude it on xlinux. I'm not sure how often it fails on the OpenJ9 machines, if ever. Or how often on the Adopt machines. Or if the xlinux failures are always timeouts.

Checked the nodes in https://ci.adoptopenjdk.net/label/ci.role.test&&hw.arch.x86&&sw.os.linux/, all have 8G RAM. The timeout from adopt is likely different, but it's not reproducible in 300 jobs.

**Comment by gita-omr** [comment]
After investigating #10580 we know that the test is trying to allocate a lot of memory. Should we change it to either pass or give an OOM exception?

**Comment by AlenBadel** [comment]
I'm not sure if it's the same issue, but JDK15 seems to hang while just doing a simple `java -Xint --version `. This has happened on non-XL PPC JDK15 builds, and was not reproducible with JDK15 XL builds, or JDK14.

This is easily reproducible by over-caching the system and pushing it into a bad state where it's starved of resources.

https://github.com/eclipse/openj9/issues/10588#issuecomment-693760224

**Comment by pshipton** [comment]
This is an IBM problem and not OpenJ9. I've opened infrastructure/issues/4797 and closing this.
